speech/src/main/java/com/example/speech/InfiniteStreamRecognize.java

```diff
@@ -293,10 +218,110 @@ public void onError(Throwable t) {}
 
           clientStream.send(request);
         }
+        clientStream.closeSend();
       } catch (Exception e) {
         System.out.println(e);
       }
     }
   }
+
+  public static ResponseObserver<StreamingRecognizeResponse> getResponseObserver() {
+    return new ResponseObserver<StreamingRecognizeResponse>() {
+
+      final ArrayList<StreamingRecognizeResponse> responses = new ArrayList<>();
+
+      public void onStart(StreamController controller) {
+        referenceToStreamController = controller;
+      }
+
+      public void onResponse(StreamingRecognizeResponse response) {
+        responses.add(response);
+        StreamingRecognitionResult result = response.getResultsList().get(0);
+        Duration resultEndTime = result.getResultEndTime();
+        resultEndTimeInMS =
+                (int) ((resultEndTime.getSeconds() * 1000) + (resultEndTime.getNanos() / 1000000));
+        double correctedTime =
+                resultEndTimeInMS - bridgingOffset + (STREAMING_LIMIT * restartCounter);
+
+        SpeechRecognitionAlternative alternative = result.getAlternativesList().get(0);
+        if (result.getIsFinal()) {
+          System.out.print(GREEN);
+          System.out.print("\033[2K\r");
+          System.out.printf(
+                  "%s: %s [confidence: %.2f]\n",
+                  convertMillisToDate(correctedTime),
+                  alternative.getTranscript(),
+                  alternative.getConfidence());
+          isFinalEndTime = resultEndTimeInMS;
+          lastTranscriptWasFinal = true;
+        } else {
+          System.out.print(RED);
+          System.out.print("\033[2K\r");
+          System.out.printf(
+                  "%s: %s", convertMillisToDate(correctedTime), alternative.getTranscript());
+          lastTranscriptWasFinal = false;
+        }
+        checkStopRecognitionFlag(alternative.getTranscript().getBytes(StandardCharsets.UTF_8));
+      }
+
+      public void onComplete() {}
+
+      public void onError(Throwable t) {}
+    };
+  }
+
+  private static void checkStopRecognitionFlag(byte[] flag) {
+    if (flag.length < 10) {
+      stopRecognition = new String(flag).trim().equalsIgnoreCase("exit");
+      putDataToSharedQueue("stop".getBytes(StandardCharsets.UTF_8));
+    }
+  }
+
+  // Microphone Input buffering
+  static class MicBuffer implements Runnable {
+    TargetDataLine targetDataLine;
+
+    public MicBuffer() throws LineUnavailableException {
+      // SampleRate:16000Hz, SampleSizeInBits: 16, Number of channels: 1, Signed: true,
+      // bigEndian: false
+      AudioFormat audioFormat = new AudioFormat(16000, 16, 1, true, false);
+      DataLine.Info targetInfo =
```

Comment:
Using hard-coded parameters in the code sample leads to flaky tests because the support for parameters can change with the time. Unless there is a specific requirement for exactly these numbers, consider to convert them to arguments passed from the main method.
If they are constants, please define them as `final` constants with meaningful names.